File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutingPolicyFactories.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.documentapi.messagebus.protocol.DocumentProtocolPoliciesConfig;
import com.yahoo.vespa.config.content.DistributionConfig;

/**
 * @author Simon Thoresen Hult
 * @author jonmv
 */
class RoutingPolicyFactories {

    private RoutingPolicyFactories() { }

    static class AndPolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new ANDPolicy(param);
        }
    }

    static class ContentPolicyFactory implements RoutingPolicyFactory {
        private final DistributionConfig distributionConfig;
        public ContentPolicyFactory(DistributionConfig config) { this.distributionConfig = config; }
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new ContentPolicy(param, distributionConfig);
        }
    }

    static class MessageTypePolicyFactory implements RoutingPolicyFactory {

        private final String configId;
        private final DocumentProtocolPoliciesConfig config;

        public MessageTypePolicyFactory(String configId, DocumentProtocolPoliciesConfig config) {
            this.configId = configId;
            this.config = config;
        }

        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            if (config != null) {
                if (config.cluster(param) == null)
                    return new ErrorPolicy("No message type config for cluster '" + param + "'");

                return new MessageTypePolicy(config.cluster(param));
            }
            return new MessageTypePolicy(param == null || param.isEmpty() ? configId : param);
        }
    }

    static class DocumentRouteSelectorPolicyFactory implements RoutingPolicyFactory {

        private final String configId;
        private final DocumentProtocolPoliciesConfig config;

        public DocumentRouteSelectorPolicyFactory(String configId, DocumentProtocolPoliciesConfig config) {
            this.configId = configId;
            this.config = config;
        }

        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            if (config != null) {
                try {
                    return new DocumentRouteSelectorPolicy(config);
                }
                catch (IllegalArgumentException e) {
                    return new ErrorPolicy(e.getMessage());
                }
            }
            DocumentRouteSelectorPolicy ret = new DocumentRouteSelectorPolicy(param == null || param.isEmpty() ?
                                                                              configId : param);
            String error = ret.getError();
            if (error != null) {
                return new ErrorPolicy(error);
            }
            return ret;
        }
    }

    static class ExternPolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            ExternPolicy ret = new ExternPolicy(param);
            String error = ret.getError();
            if (error != null) {
                return new ErrorPolicy(error);
            }
            return ret;
        }
    }

    static class LocalServicePolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new LocalServicePolicy(param);
        }
    }

    static class RoundRobinPolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new RoundRobinPolicy();
        }
    }

    static class LoadBalancerPolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new LoadBalancerPolicy(param);
        }
    }

    static class SubsetServicePolicyFactory implements RoutingPolicyFactory {
        public DocumentProtocolRoutingPolicy createPolicy(String param) {
            return new SubsetServicePolicy(param);
        }
    }

}

File: documentapi/src/vespa/documentapi/messagebus/routingpolicyfactories.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "routingpolicyfactories.h"
#include <vespa/documentapi/messagebus/policies/andpolicy.h>
#include <vespa/documentapi/messagebus/policies/contentpolicy.h>
#include <vespa/documentapi/messagebus/policies/documentrouteselectorpolicy.h>
#include <vespa/documentapi/messagebus/policies/errorpolicy.h>
#include <vespa/documentapi/messagebus/policies/externpolicy.h>
#include <vespa/documentapi/messagebus/policies/loadbalancerpolicy.h>
#include <vespa/documentapi/messagebus/policies/localservicepolicy.h>
#include <vespa/documentapi/messagebus/policies/messagetypepolicy.h>
#include <vespa/documentapi/messagebus/policies/roundrobinpolicy.h>
#include <vespa/documentapi/messagebus/policies/subsetservicepolicy.h>
#include <vespa/config/subscription/configuri.h>

using namespace documentapi;

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::AndPolicyFactory::createPolicy(const string &param) const
{
    return std::make_unique<ANDPolicy>(param);
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::MessageTypePolicyFactory::createPolicy(const string &param) const
{
    return std::make_unique<MessageTypePolicy>(config::ConfigUri(param));
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::ContentPolicyFactory::createPolicy(const string &param) const
{
    auto ret = std::make_unique<ContentPolicy>(param);
    string error = static_cast<ContentPolicy&>(*ret).getError();
    if (!error.empty()) {
        return std::make_unique<ErrorPolicy>(error);
    }
    return ret;
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::LoadBalancerPolicyFactory::createPolicy(const string &param) const
{
    mbus::IRoutingPolicy::UP ret(new LoadBalancerPolicy(param));
    string error = static_cast<LoadBalancerPolicy&>(*ret).getError();
    if (!error.empty()) {
        fprintf(stderr, "Got error %s\n", error.c_str());
        return std::make_unique<ErrorPolicy>(error);
    }
    return ret;
}

RoutingPolicyFactories::DocumentRouteSelectorPolicyFactory::
DocumentRouteSelectorPolicyFactory(const document::DocumentTypeRepo &repo,
                                   const string &configId) :
    _repo(repo),
    _configId(configId)
{
    // empty
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::DocumentRouteSelectorPolicyFactory::createPolicy(const string &param) const
{
    auto ret = std::make_unique<DocumentRouteSelectorPolicy>(_repo, config::ConfigUri(param.empty() ? _configId : param));
    string error = static_cast<DocumentRouteSelectorPolicy&>(*ret).getError();
    if (!error.empty()) {
        return std::make_unique<ErrorPolicy>(error);
    }
    return ret;
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::ExternPolicyFactory::createPolicy(const string &param) const
{
    auto ret = std::make_unique<ExternPolicy>(param);
    string error = static_cast<ExternPolicy&>(*ret).getError();
    if (!error.empty()) {
        return std::make_unique<ErrorPolicy>(error);
    }
    return ret;
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::LocalServicePolicyFactory::createPolicy(const string &param) const
{
    return std::make_unique<LocalServicePolicy>(param);
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::RoundRobinPolicyFactory::createPolicy(const string &param) const
{
    return std::make_unique<RoundRobinPolicy>(param);
}

mbus::IRoutingPolicy::UP
RoutingPolicyFactories::SubsetServicePolicyFactory::createPolicy(const string &param) const
{
    return std::make_unique<SubsetServicePolicy>(param);
}
